### ISSUE TITLE
Feature request: add one hand mode

### ISSUE THREAD

**Issue by akfish** [philc/vimium#1455]
## Motivations

It is often for a user wanting/having to use only one of his/her/its hand, situations like:
* He/she/it is too hungry to put both hands on the keyboard due to lack of food
* He/she/it is currently holding food with one hand
* One of his/her/its hand is dirty because he/she/it was just holding food

It would be useful if Vimium can show link hints with letters only on one side of the keyboard so that a user, should he/she/it chooses to, can eat happily in front of the computer all the time, get fat and die of heart disease earlier than expected. :)

## Possible Solutions

1. Display alternatives for both hands at the same time
2. Add hot key to switch mode. `A` for left-hand mode, `;` for right-hand mode and `h` for both. It should be very intuitive if you look at those keys' position.

**Comment by z0rch** [comment]
@akfish, have a look at #1421. Perhaps you could provide your input there as well.

**Comment by akfish** [comment]
@z0rch Hmm. An exact opposite feature request. Interesting.

**Comment by gdh1995** [comment]
Um...
Food is only an example,
and in most cases my right hand is holding my mouse, writing something, and so on.

**Comment by gdh1995** [comment]
I think, what we need is just a status:
* meaning which keyboard layout is NOW wanted;
* can be switched to another `conveniently: less keys used; called by content scripts`
* persistent: should be passed to other tabs and frames whenever it's changed (easy to implement);

Although `A, h, ;` is intuitive, it's not a perfect choice, unless:
* there's a way to tell these controller keys from normal hints' keys

If we can ?
* accept "L" as a controller when in a left-hand mode (or a full one)
* "A" in a right-hand mode (or a full one)
* someone like "H" in left/right-hand modes
Or one key to control if in the full mode and another to control which hand?

**Comment by akfish** [comment]
@gdh1995 Yep. I was hungry then and food was all I could think of.

As for the hotkeys, I don't see any problems. Mode switch happened before hints are displayed (before 'F'), like any other hotkeys. So hints selection won't be messed up. The interaction flow is like:
1. The user is holding a hotdog in his right hand and reaches for keyboard with the other
2. The user presses A. Vimium switches to left hand mode
3. The user presses F. Vimium displays link hints generated in left hand mode
4. The user navigates with hints while eating his hotdog

**Comment by mrmr1993** [comment]
This is a really cool feature idea. I've had this problem in the past, it would be really nice to have a way around it! Entering the keys for each hand in 2 boxes and implementing this in parallel with #1421 makes the most sense to me.

For modifiers, I would suggest the function keys, using their position to describe the mode:
* `F1` - left hand only
* `F6` - default
* `F7` - alternating hands (as in #1421)
* `F12` - right hand
* `F2`-`F5`, `F8`-`F12` - (possibly) dead keys, so if a user accidentally changes mode they can just mash the function keys until the find the right one to put it back.

This benefits from being easy to add into the current code, having no localisation issues, and leaving all printing keys available for link hints. The obvious annoyances would be that you can't swap to the Omnibox or open developer tools while link hints are showing.

Changing the hint style only when `F` had been pressed but no link hints have been filtered is the simplest, most self-contained way I can think of. Also, then the user can see the effects of the mode change immediately, which is helpful.

**Comment by akfish** [comment]
@mrmr1993 Thanks for the support. 
The problem with function keys is that they are usually used by browser itself (e.g. `F1` for help, `F3` for search, `F12` for dev tools and etc).

**Comment by mrmr1993** [comment]
>The problem with function keys is that they are usually used by browser itself

Is there a compelling use case for any of the browser's function key actions while link hints are active? The rest of the time they should definitely be left alone, but (apart from developers using `F12` to examine the link hints &mdash; which could easily be done before activating them) I can't see any reason why a user would have link hints active and then use one of those actions.

**Comment by akfish** [comment]
Well, it's not about that. 

It is generally not a good practice to mess with browsers default behaviors, especially when they are not intended to be messed with. Reasons I could think of are:
1. It may yield unpredictable/inconsistent result in different platforms.
2. It is undocumented that how the browser should behave when you try to override its default behavior. Thus it may be subject to change in the future.
3. It may confuse users since it's against their long-learned knowledge

Although you can test them all, document it yourself and implement it anyway. Then one day the browser develop team wake up and decide to screw your efforts in a permanent way. It would be all for nothing.

Anyway, users can use whatever hotkeys they like since key mapping is configurable. I just don't think it would be wise to mess with the browser by default.

**Comment by gdh1995** [comment]
About `F1`-`F12`, :
* I hate `F1` very much - `support.google.com` is nothing helpful ;
* `F12` is so useful to test code changes, that it should be passed everywhere;
* `F3` is not very useful when `LinkHints` is active, but we can't say it will never be used;
* As for messing with default behaviors, it seems a few of them, such as `F1` and `F2`, can be reused. 
(I've prevented `F1` and released `F12` in my customized Vimium)

Okay, all above belong to another issue "How to use `F1`-`F12`". If we choose "press something before entering the LinkHints mode", just a common sequence of keys is OK.
So, what about `Fa` for left, `Fl` for right, ... Then we need to add 3 commands and map them.

**Comment by mrmr1993** [comment]
> It is generally not a good practice to mess with browser's default behaviors, especially when they are not intended to be messed with.

@akfish The [spec](http://www.w3.org/TR/DOM-Level-3-Events/#event-type-keydown) lays out pretty straightforwardly that we *are* intended to mess with them if we so choose, and goes about detailing the ways in which the browser should allow us to do so. There is nothing wrong with us using the function keys *unless we are concerned that doing so will confuse or disrupt the user*.
Clearly we would not be, as you can test yourself by trying to use a function key in link hint mode: you'll see that, not only are we already overriding the default action, but that nobody has noticed to date.

> It may yield unpredictable/inconsistent result in different platforms.

The function keys are actually massively more consistent than letter keys with their `keycode`s. #1387, #1411, #1417 and #576 give an idea of the problems with the non-function keys. This will be a problem with any keys we try to handle until [Chromium issue 227231](https://code.google.com/p/chromium/issues/detail?id=227231) is fixed.

> It is undocumented that how the browser should behave when you try to override its default behavior. Thus it may be subject to change in the future.

[Spec](http://www.w3.org/TR/DOM-Level-3-Events/#event-type-keydown) again.

> It may confuse users since it's against their long-learned knowledge

This is the only real issue. As I mentioned before, function keys haven't been working in link hint mode for quite some time and nobody noticed, so it doesn't look like this would be a problem.

**Comment by akfish** [comment]
Thanks for the information. But actually, the spec only specifies the event system **inside** DOM. Browsers and OS level keyboard event handling is completely out of its scope and should take precedence over it.

And the spec only said that you **can** do such a thing. But it didn't say it is **recommended** or whether it is a good idea to do so. Specs only tell you what infrastructure will/should be provided. How you should use it is usually a different story, and in this case is dictated by HCI principles.

Another reason I didn't mention last time is the consistence in logic. The whole point (or advantages) of Vim-like key mapping scheme is for users not to have to move their figures away from the ready position often. And the reason for requesting single hand mode requires that even more since users' ability to use keyboards are assumed to be impaired by some obstacles (a.k.a the food). Function keys are relatively further away and (sometimes extremely) hard to locate them correctly, especially on compact keyboards like laptops':
![Very Narrow Keyboards](https://cloud.githubusercontent.com/assets/922715/6054967/a4f456b0-ad30-11e4-9765-4a0f7389a241.png)
I could tell you even with 2 hands it's not easy. Imagine try that with one hand while eating.

Well, I think we're a little carried away here. Key mappings are minor issue here and I'm not trying to  convince you or something. Let's call it an exchange of ideas.

**Comment by gdh1995** [comment]
Well, it's an important fact: we "seldom" need to change hint's key sequence, don't we?
The function "enable one hand mode" is not used as frequently as those such as `scrollDown`, and `nextTab`.
Then why not choose a key not at hand? So that more letter keys are free to other commands.

**Comment by akfish** [comment]
Note that this feature is designed for situations where users' capabilities
are compromised. A different mindset should be used when designing things.
I know it's a little overstatement. But you wouldn't expect blind people to
use the computer as normal ones do, wouldn't you?

When a user realizes that he needs to switch one hand mode, it is usually
"too late": He is eating. He puts his hands on the keyboard. And oops,
can't use the other hand. He reaches for function keys, which is hard to
hit correctly even with 2 hands and no distractions. That's not much better
(if not even worse considering an extra step is taken) than the vanilla two
hands mode.
Now let's try that again with a different key mapping: he is eating with
his right hand. He puts his left hand on the keyboard. And oops, can't use
the right hand. But it's OK. Left hand is in position anyway. He hits `A`
comfortably. And from then on, every keystroke is in his reach. Mission
accomplished.

**Comment by gdh1995** [comment]
So, can we use `a` for left-hand LinkHints, `l` for right-hand LinkHints, and `f` for the default? This seems ugly and uncomfortable, but I have no more idea...
If I'm eating, my left hand is glad to press a function key like `f1`, for this key should be pressed only once - not a tiring work

**Comment by akfish** [comment]
Well, I guess the real options from here on is to do some prototyping and
find out. Unfortunately I'm occupied by other projects. If anyone is
interested in picking up this issue, I'm glad to help with the test
process. If that is not happening, I will revisit this and send a PL of my
own once I have the time.

**Comment by spelufo** [comment]
I set up some hotkeys and use my left hand only most of the time. They might override some stuff I didn't want to use. Here they are, in case someone wants them.

```
map e scrollUp
map d scrollDown
map s nextTab
map S moveTabRight
map a previousTab
map A moveTabLeft
map z goBack
map c goForward
map v Vomnibar.activateTabSelection
map t Vomnibar.activateInNewTab
```

**Comment by dragon788** [comment]
Just my 2 cents, but if I'm typing with one hand, I'm not quite as concerned about "touch typing" but I do hope that there are no shortcuts that cause me injury while trying to hit them with one hand. I believe the function buttons would be fine for most cases, but on a Mac where the default is a system change (brightness/volume/etc) hitting Fn + F# isn't feasible with one hand.

**Comment by rrsahu** [comment]
When you are working with us, we ask you to fill a simple form so that we can get an concept about your specifications. Once we are sure about it, we will record you best Company information within few minutes only. 

<a href="http://www.expert5th.in/packers-and-movers-hyderabad/" title="Packers and Movers Hyderabad">Packers and Movers Hyderabad</a>
<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers Bangalore">Packers and Movers Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers Mumbai">Packers and Movers Mumbai</a>

**Comment by rrsahu** [comment]
Quotations are usually available from all major Companies and it is simple to get concept about their solutions. 

<a href="http://www.expert5th.in/packers-and-movers-gurgaon/" title="Packers and Movers Gurgaon">Packers and Movers Gurgaon</a>
<a href="http://www.expert5th.in/packers-and-movers-delhi/" title="Packers and Movers Delhi">Packers and Movers Delhi</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers Pune">Packers and Movers Pune</a>

**Comment by rrsahu** [comment]
These Information mill not only providing product packaging and movers in Bangalore but other significant places as well.

<a href="http://www.expert5th.in/packers-and-movers-chennai/" title="Packers and Movers Chennai">Packers and Movers Chennai</a>
<a href="http://www.expert5th.in/packers-and-movers-noida/" title="Packers and Movers Noida">Packers and Movers Noida</a>
<a href="http://www.expert5th.in/packers-and-movers-navimumbai/" title="Packers and Movers Navi Mumbai">Packers and Movers Navi Mumbai</a>
<a href="http://www.expert5th.in/packers-and-movers-thane/" title="Packers and Movers Thane">Packers and Movers Thane</a>
<a href="http://www.expert5th.in/packers-and-movers-ghaziabad/" title="Packers and Movers Ghaziabad">Packers and Movers Ghaziabad</a>
<a href="http://www.expert5th.in/packers-and-movers-faridabad/" title="Packers and Movers Faridabad">Packers and Movers Faridabad</a>

**Comment by rrsahu** [comment]
You have been got an concept about solutions and packages; next step is to create booking with most reliable Company as per your moving need and specifications. 

<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers in Bangalore">Packers and Movers in Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers in Pune">Packers and Movers in Pune</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers in Mumbai">Packers and Movers in Mumbai</a>

**Comment by keyan** [comment]
I was just eating a croissant and trying to click on a link with vimium when I thought "enough is enough, how can I get the link shortcuts to only use keys on the left-hand side?". And now I am here. For context, I use a [kinesis advantage keyboard](https://kinesis-ergo.com/wp-content/uploads/kb600-angled-l-2007x1024.jpg) so it is especially hard to move one hand between sides.

Just throwing in some support for this feature.

**Comment by v3ai** [comment]
I also have interest here I want this feature
